### PR TITLE
PEP updates only no xlinke

### PR DESCRIPTION
**Pull request by trishorts**
some updates that will make pep run more smoothly

### CODE REVIEW

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 1]
```diff
@@ -240,17 +280,18 @@ private static float GetSSRCalcHydrophobicityZScore(PeptideSpectralMatch psm, Pe
...
+                int time = (int)(2 * Math.Round(psm.ScanRetentionTime / 2d, 0));
                 if (d[psm.FullFilePath].Keys.Contains(time))
                 {
                     double predictedHydrophobicity = calc.ScoreSequence(Peptide);
                     hydrophobicityZscore = Math.Abs(d[psm.FullFilePath][time].Item1 - predictedHydrophobicity) / d[psm.FullFilePath][time].Item2;
                 }
             }
 
-            if (double.IsNaN(hydrophobicityZscore) || double.IsInfinity(hydrophobicityZscore))
+            double maxHydrophobicityZscore = 10;
```
Comment (zrolfs) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: Can you comment on why this is 10? Instead of, say, 100? If it's arbitrary, note it.

Comment (trishorts) [reply EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: 1 Z is one standard deviation. 10 standard deviations is lots.

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 2]
```diff
@@ -170,14 +170,21 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
...
                     foreach ((int notch, PeptideWithSetModifications pwsm) in psm.BestMatchingPeptides)
                     {
-                        if (pwsm.AllModsOneIsNterminus.Any() && !computeHydrophobicitiesforModifiedPeptides)
+                        if (fullSequences.Contains(pwsm.FullSequence))
+                        {
+                            continue;
+                        }
+                        fullSequences.Add(pwsm.FullSequence);
+                        double predictedHydrophobicity = calc.ScoreSequence(pwsm);
+                        int possibleKey = (int)(2 * Math.Round(psm.ScanRetentionTime / 2d, 0));
```
Comment (lonelu) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: int possibleKey = (int)(2 * Math.Round(psm.ScanRetentionTime / 2d, 0)):
Could you add annotation/explain?

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 3]
```diff
@@ -332,6 +373,9 @@ public static PsmData CreateOnePsmDataEntry(PeptideSpectralMatch psm, Dictionary
             if (trainingVariables.Contains("PsmCount"))
             {
                 psmCount = sequenceToPsmCount[String.Join("|", psm.BestMatchingPeptides.Select(p => p.Peptide.FullSequence).ToList())];
+                List<int> psmCountList = new List<int> { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 30, 40, 50, 75, 100, 200, 300, 400, 500 };
```
Comment (lonelu) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: I believe you have a reason why divide it using the psmCountList. Can you explain?

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 4]
```diff
@@ -203,10 +211,45 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
...
+                            allStDevs.Add(Math.Pow(difference, 2));
+                        }
+                    }
+                }
+
+                //some standard deviations are too small or too large because of random reasons, so we replace those small numbers of oddballs with reasonable numbers.
+                double globalStDev = 1;
+                if (allStDevs.Count() > 1)
+                {
+                    globalStDev = Math.Sqrt(allStDevs.Average());
```
Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: globalHydrophobicityDifference?

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 5]
```diff
@@ -170,14 +170,23 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
                 Dictionary<int, List<double>> hydrobophobicites = new Dictionary<int, List<double>>();
```
Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: hydrophobicites (spelling)

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 6]
```diff
@@ -170,14 +170,23 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
...
                 Dictionary<int, Tuple<double, double>> averagesCommaStandardDeviations = new Dictionary<int, Tuple<double, double>>();
```
Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: It would be good to change this in a later PR from the `Tuple<double,double>` structure to something like `(double name1, double name2)` so that it's a bit easier to follow the variable names. this code is pretty hard to follow

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 7]
```diff
@@ -203,10 +211,45 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
                     }
                 }
 
+                List<double> allStDevs = new List<double>();
+
                 foreach (int key in hydrobophobicites.Keys)
```
Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: `roundedRtBin` or something similar instead of `key` is more clear

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 8]
```diff
@@ -203,10 +211,45 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
...
+                {
+                    globalStDev = Math.Sqrt(allStDevs.Average());
+                }
+
+                Dictionary<int, Tuple<double, double>> stDevsToChange = new Dictionary<int, Tuple<double, double>>();
+                foreach (KeyValuePair<int, Tuple<double, double>> item in averagesCommaStandardDeviations)
+                {
+                    //add stability. not allowing stdevs that are too small or too large at one position relative to the global stdev
+                    //here we are finding which stdevs are out of whack.
+                    if (Double.IsNaN(item.Value.Item2) || item.Value.Item2 < 0.5 || (item.Value.Item2 / globalStDev) > 3)
```
Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: I don't understand the `item.Value.Item2 < 0.5` criterion. is it bad to have a stddev of hydrophobicity differences below 0.5?

Comment (rmillikin) [inline EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs]: I don't think globalStDev is being calculated properly. line 235 is the global difference in hydrophobicity and bin hydrophobicity; I don't think it's been z-transformed
